### DISCUSSION TITLE
How to use `:trust`?

### DISCUSSION THREAD

**Discussion by HiPhish** [neovim/neovim#27627] [Q&A] [answered]
Hello, I'm trying to automate trusting a file (`:h vim.secure` and `:h :trust`) for a test case. I tried manually executing the following:

```vim
" Use a relative path
trust .busted

" Use absolute path
trust ~/Developer/vim/start/neotest-busted/.busted
```

But each time I get `E475: Invalid argument: ~/Developer/vim/start/neotest-busted/.busted` or whatever the path is. I tried putting quotes around the path but I get the same error. Is there something I am missing here?

I can trust a file when I allow it manually after calling `vim.secure.read`, but I want to whitelist the file without an interactive prompt because I want to use it in automated testing. For some reason `vim.secure.trust` does not allow file paths with `action = 'allow'` (why though?), so the `:trust` command seems like the way to go.

My Neovim version is 0.9.5 on [Void Linux](https://voidlinux.org/) installed from their package repos.

**Comment by zeertzjq** [comment]
You can only trust the current file. Trusting non-current file is not allowed, for security reasons (e.g. a malicious process may change its content behind your back while you are typing the command).

**Reply by HiPhish** [reply]
Is this really a realistic attack vector? A plugin which can whitelist a file can already do much more damage with the `system` or `jobstart` functions. Exploiting the trust system is a really roundabout way to attack someone when there are much more direct attacks possible.

If we are talking about an external process, it could just edit the `trust` file directly to inject a new entry. It would not have to go through Neovim first.

Where I see the point of the trust system is when it comes to Neovim executing code which you have not explicitly added to your config. A simple example, you might have Neovim set up to load project-specific settings from a local `.vimrc` file, so an attacker could put such a file in a repository, and when you open some other file from the repo you end up sourcing the unknown `.vimrc` file.

**Reply by gpanders** [reply]
>Is this really a realistic attack vector?

[Yes](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use).

**Reply by HiPhish** [reply]
@gpanders I still do not understand how that type of attack applies here. Here is a potential attack scenario I can imagine:

1. I download a project with a file `.nvimrc` which I do in fact know is safe
2. I open Neovim and manually start typing out something like `lua vim.secure.trust('.nvimrc')`
3. In the meantime a malicious process inserts some bad code into `.nvimrc`
4. By the time I press `<CR>` the file has been modified and I have now trusted and executed bad code

The weakness I see here is that an attacker could circumvent the trust check:

1. As before
2. A malicious process manipulates `.nvimrc`, computes its hash and inserts a new entry into the trust database
3. I open `.nvimrc` expecting to get asked whether I am willing to trust it, but no such question is shown

At this point I have already executed to malicious code. This looks like barricading the front door while leaving the back door open.

**Reply by brianhuster** [reply]
That's why it calculates hash of the _content_ in your buffer

**Reply by HiPhish** [reply]
I can compute the hash of a file on disc:

```lua
local function to_truststring(file)
	local fullpath = vim.fn.fnamemodify(file, ':p')
	local f = io.open(fullpath, 'r')
	if not f then error('Cannot read file ' .. file) end
	local contents = f:read('*a')
	f:close()
	local hash = vim.fn.sha256(contents)
	return string.format('%s %s', hash, fullpath)
end

local trustfile = vim.fn.stdpath('state') .. '/trust'
local trust_content = vim.tbl_map(to_truststring, trusted_files)
vim.fn.writefile(trust_content, trustfile, 's')
```

In fact, this is precisely what I am doing as a workaround in [hiphish/neotest-busted](https://github.com/hiphish/neotest-busted) to create an artificial trust file for tests. Here I am using Neovim's built-in Lua interpreter, but an attacker could use any other SHA256 implementation for this trick.

**Reply by justinmk** [reply, marked as answer]
> I can compute the hash of a file on disc:

But the point of the restriction is that you view the contents in a buffer. That protects against a malicious process changing the on-disk contents.

Nevertheless, with https://github.com/neovim/neovim/pull/38001 `:trust [path]` is allowed now.
